packages/squiggle-lang/src/reducer/index.ts

```diff
@@ -249,24 +250,29 @@ const evaluateLambda: SubReducerFn<"Lambda"> = (
   return [value, context];
 };
 
-const evaluateCall: SubReducerFn<"Call"> = (expressionValue, context, ast) => {
+const evaluateCall: SubReducerFn<"Call"> = (
+  expressionValue: Extract<Expression, { type: "Call" | "Decorate" }>["value"],
+  context: Context.ReducerContext,
+  ast: ASTNode
+) => {
   const [lambda] = context.evaluate(expressionValue.fn, context);
+  if (lambda.type !== "Lambda") {
+    throwFrom(new RENotAFunction(lambda.toString()), context, ast);
+  }
+  if (expressionValue.as === "decorate" && !lambda.value.isDecorator) {
```

Comment:
This is an extra check that we'll have to do on each function call, but the performance overhead is not noticeable.